File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.java
import java.util.*;

class Solution {
    public int solution(int n, int k, int[] enemy) {
        PriorityQueue<Integer> qu = new PriorityQueue<>(Collections.reverseOrder());

        for (int i = 0; i < enemy.length; i++) {
            int e = enemy[i];

            qu.offer(e);
            n -= e;

            if (n < 0) {
                if (k > 0) {
                    n += qu.poll();
                    k--;
                } else {
                    return i;
                }
            }
        }

        return enemy.length;
    }
}

File: 프로그래머스/2/142085. 디펜스 게임/디펜스 게임.py
import heapq

def solution(n, k, enemy):
    qu = []
    
    for i, e in enumerate(enemy):
        heapq.heappush(qu, -e)
        n -= e
        
        if n < 0:
            if k > 0:
                n += -heapq.heappop(qu)
                k -= 1
            else:
                return i
    
    return len(enemy)